module/zfs/zio_compress.c

```diff
@@ -143,8 +143,18 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len,
 	if (c == ZIO_COMPRESS_EMPTY)
 		return (s_len);
 
-	/* Compress at least 12.5% */
-	d_len = s_len - (s_len >> 3);
+	/*
+	 * Usable compression thresholds are:
+	 * - less than BPE_PAYLOAD_SIZE (embedded_data feature)
+	 * - at least one saved sector
+	 */
+	if (compress_threshold > 0) {
+		d_len = MAX(BPE_PAYLOAD_SIZE,
+		    s_len - (s_len % compress_threshold));
+	} else {
+		/* Special case for reproducibility of ARC checks */
+		d_len = s_len;
+	}
 
```

Comment:
I think we should `return (s_len);` right here if `d_len >= s_len` (which might now be the case if, say, `BPE_PAYLOAD_SIZE >= s_len` because although it won't waste any (more) physical space than avoiding compression, it's going to be a waste of CPU to go ahead with compression.

Reply:
(or, uh, just check that s_len < BPE_PAYLOAD_SIZE back in the `if (compress_threshold > 0) {` path...)

Reply:
ignore, brainfart